app/controllers/oidc_users_controller.rb

```diff
@@ -6,12 +6,35 @@ def update
     user.set_local_attributes(params.permit(OIDC_USER_ATTRIBUTES).to_h)
     attributes = user.get_local_attributes(OIDC_USER_ATTRIBUTES)
 
-    if attributes["email"] && !attributes["email_verified"].nil?
-      user.email_subscriptions.each do |email_subscription|
-        email_subscription.reactivate_if_confirmed!(
-          attributes["email"],
-          attributes["email_verified"],
+    if attributes["email"] && attributes["email_verified"]
+      begin
+        # if the user has linked their notifications account to their
+        # GOV.UK account we don't need to update their
+        # `user.email_subscriptions`, because we can update the
+        # subscriber directly.
+        subscriber_id = GdsApi.email_alert_api
+          .find_subscriber_by_govuk_account(govuk_account_id: user.id)
+          .dig("subscriber", "id")
+        GdsApi.email_alert_api.change_subscriber(
+          id: subscriber_id,
+          new_address: attributes["email"],
         )
+      rescue GdsApi::HTTPNotFound
+        # but for users who haven't linked their notifications account
```

Comment:
This seems like quite the oversight.
Are we sure that wasn't deliberate?

I can't see why it would be though 🤔 ...
This comes straight after a notification subscriber has proven access to an email account right?
So there aren't any authentication benefits... we'd just be asking the user to repeat the process N times...

And if folks had a new email address or wanted to only update some... they could just create a new (perhaps a "work") account.

So this looks good to me!

Reply:
I wonder how many subscriptions the average user has now and how frequently they update?
Any chance we'll need more workers if this becomes a regular operation?

Reply:
I'm not sure what you mean, what oversight?  This endpoint is called when a user changes the email address of their GOV.UK Account, not their Notifications Account.

There are 3 possible states a user could be in:

1. A user has a GOV.UK Account and a Notifications Account and they are linked
2. A user has a GOV.UK Account and a Notifications Account and they are not linked
3. A user has a GOV.UK Account and no Notifications Account

So, we call email-alert-api to see if there is a Notifications Account linked to this GOV.UK Account.  If there is, we're in case 1 and can update the Notifications Account directly.

If there isn't, we're in either case 2 or 3.  And in both cases the behaviour is the same: update the address of only the email subscriptions which account-api knows about (eg, the Transition Checker subscription), because the expected behaviour of those subscriptions is that they are linked to the account.

account-api doesn't have a database of all subscriptions, only those which are explicitly connected to the account.  Which is currently just the Transition Checker subscription, but in the future may also include others.

Reply:
> But any other subscriptions for the same email address won't. This isn't great.

ah ok sorry I thought this meant that the current behaviour of the notifications was that an unlinked Notifications account can only update the email address on a per-subscription list basis?

So the commit message isn't making that comparison then?

Reply:
Individual subscriptions don't have an email address associated with them, they belong to a Notifications Account, and a Notifications Account has an email address.

So the current behaviour for unlinked accounts where we do nevertheless have an individual subscription which account-api knows about (like the brexit checker subscriptions), is to delete the old one and make a new one for the new email address - recreating the subscription under a *different* Notifications Account*.

But when a Notifications Account and a GOV.UK Account are linked, we don't need to move subscriptions between Notifications Accounts, we can just update the Notifications Account itself.  Does that help?